### ISSUE TITLE
Multi-file upload: hard to find the uploaded files

### ISSUE THREAD

**Issue by phiw13** [textpattern/textpattern#938]
One UI issue with the [multi-file upload feature](https://github.com/textpattern/textpattern/commit/b106a0741ce2b0c5a7d71d1650a9845235d1b695) is that it makes it hard for the user to immediately “see” the uploaded files.

Example: on my local playground test site, I had the files panel sorted by category (4 pages worth of it), with the category-less files at the end of the list. I uploaded 3 files, the message pane told me the upload was successful (but warning! one file already exists - that is OK), but then I have to hunt for the files. If they appear on the first page, they’re at least highlighted, but if they’re on a following page, they’re harder to track.

I don’t have any good idea to solve that, maybe (not great… stomping on user setting) sort the files by ID?

(the issue is not really new, as it exist sort-of on TXP 4.6x and older, but there after uploading the file, the Edit panel would pop open allowing the user to fill in meta data, categories and what not).

**Comment by philwareham** [comment]
Hmm, that's a valid issue. I wonder if there is a way to then go to the edit panel for each upload sequentially after upload?

**Comment by bloatware** [comment]
Sorting freshly uploaded files by id does not help when their number is larger than `pageby` value. We could, though, limit this number by `pageby`.

Multiple files edit panel should be doable, if our UX department finds it reasonable, but tricky. I would avoid it :-)

**Comment by philwareham** [comment]
Thinking out loud again...

Maybe if a user drops a single file into the uploader (probably the most common way files will be uploaded) then it uploads and invokes the edit panel. If a user uploads a batch of files then it just inserts them into the list/database (as it does now) and it's up to the user to find, select and edit.

It's still better than the old UI either way - and we get progress feedback as the file uploads too which is nice (in the old UI if a large file is uploading users have no idea if it's happening or not until it has finished upload).

I need to explore the widget more to see all it can do. Lots of config options available.

**Comment by bloatware** [comment]
That's exactly like it works now... but without Dropzone.js. We still need to discover how to integrate it seamlessly.

Purely in core, we could add new file edit links to the message pane.

**Comment by Bloke** [comment]
Haven't tried this yet but I like the direction it's heading!

Adding edit buttons alongside each uploaded file in the progress box would be neat, as long as people could middle-click to open them in different tabs to not lose the uploaded list.

One other thing: does this branch ease the burden by having a category selector near the Upload widget? So all files could all automatically go in the nominated category? If not, could it gain that ability?

Other meta data like Title is harder to extract and would probably need to be manually assigned (unless we dumbed-up the filename and converted hyphens/underscores to spaces?!)

**Comment by philwareham** [comment]
We should be able to do some if not most of what you need. The HTML behind the uploader is totally customisable from what I've read so far.

I've been starting on styling the widget (it's a work-in-progress) [here](http://design-patterns.textpattern.io/mockups/content-image.html). I think the widget should appear if you click on a 'Upload images' button exactly how it does in WordPress. Keeps clutter to a minimum and maintains layout and UI consistency with all other list panels. Still playing with layouts so this may change.

**Comment by bloatware** [comment]
I have already added a quick-n-dirty category selector, just not committed yet. But except this point, things are different atm for `dev` and `dropzone` branches, so let's be clear here.

* dev: one can upload multiple files by clicking "Upload" button. This makes a normal post, updating db, the files list and the message pane (with edit links).
* dropzone: from what I've read, it is impossible to post files via non-ajax request, so we stay on the same page. Which currently means we do not know the new files ids and they don't appear on the list until redirection. But on redirection we do not know anymore which files have just been uploaded. We could though add `Title` input *before* uploading.

Everything is solvable, we need just keep in mind that Dropzone.js is *not core* but *theme* related thing. Another theme could use plupload or nothing at all. So please don't use propitiatory markup in core elements.

**Comment by philwareham** [comment]
Would plupload be a better choice in this respect? Don't want to go too far with any CSS if it turns out this widget doesn't work as people want it too.

**Comment by bloatware** [comment]
Dunno for sure, but the problems must be the same. Our current upload process is not ajax, so we'd need to change it anyway to get a decent result.

**Comment by Bloke** [comment]
You can definitely return the IDs from plupload via the JSON object it returns. I've done that before, but I don't recall the specifics.

**Comment by bloatware** [comment]
I know nothing about upload, but how would any third-party script know the ids of txp db since we don't send them (yet)? Does it include a server script?

**Comment by Bloke** [comment]
Yep. You write an `upload.php` component, which you then tell the client side to run when upload is initiated. That can do whatever it wants and returns a JSON block, which other callbacks latch onto and disperse the returned info throughout the application as we see fit.

The only tricky part of the AJAX stuff is securing it in case someone decides to hack it, or call one of our other server-side PHP files, since the destination `upload.php` is (I believe) named in the JS object, which is visible to all and can thus be manipulated in the payload. I think we're covered with other files, but if we're opening up AJAX in this way, it's something to be wary of.

**Comment by bloatware** [comment]
Please test [https://github.com/textpattern/textpattern/commit/52dd40ca0d2bb306e5a27c3f3e1e5dfd462275d4](https://github.com/textpattern/textpattern/commit/52dd40ca0d2bb306e5a27c3f3e1e5dfd462275d4) for the functionality.

I don't think we should adapt core to please upload libraries, it's other way :-)

**Comment by bloatware** [comment]
At [present](https://github.com/textpattern/textpattern/commit/da7a301aa2e5de7e22a5b9210b93382a6ddb8848), the uploaded files boxes are checked, and the selection is persistent while filtering/sorting/navigating the file list. Does this close the issue?

**Comment by phiw13** [comment]
Combined with the categories select widget that you’ve added, this is already a good improvement. If other people are happy, then this can be closed.

**Comment by bloatware** [comment]
I mean "hard to find the uploaded files" part, the upload widget itself still needs some love.

**Comment by bloatware** [comment]
@Bloke Basic multiple files/images upload transport can actually be shipped with 4.7 as expected. We then could enhance it with more elaborated client-side features (image resize/thumbnails etc) in 4.8. Here, the delay is justified by the lack of flexible form widgets (both server and client side) you evoked in another thread.

**Comment by philwareham** [comment]
I need to work out what needs to be styled up here. Not sure what scripts are being used now (BlueImp? Dropzone?) so I don't know what CSS I need to hook up to. Any explanations would help.

FYI this was initially pushed to 4.8 during our weekend meeting because of changes we hope to have in the images handling at that version, which might have a bearing on this (or might not) and the timeframe in which we wish to release 4.7. However if you are happy with this current implementation then I could squeeze it into 4.7 depending on CSS work needed.

I would advise that we don't have any image manipulation or any images displaying during upload for now, just a simple progress bar, if that's possible? I still need to check out this branch and test more.

**Comment by bloatware** [comment]
> I would advise that we don't have any image manipulation or any images displaying during upload for now, just a simple progress bar, if that's possible?

Yes, sure, we can postpone extra UX elements to 4.8, exposing only a OS progress bar and a category selector in 4.7. FWIW, we are using BlueImp for chunked files upload (to be able to upload oversized files), but without its user interface, so you don't really need to delve into it atm. If you have time to properly style the `div` on line 2020 of the current `textpattern.js`, I guess we can also ship a basic preview with a `title` field. Please test.

**Comment by philwareham** [comment]
Cheers Oleg, thanks. I was in no way negating the work you've done here. It is great and very welcome (and very needed). Just time is against us for shipping 4.7!

I've also pushed my image grid layout back to 4.8 release, for the same reasons.

**Comment by phiw13** [comment]
Just fwiw, I had made a [WIP-attempt at styling a preview box]:https://dev.l-c-n.com/txp/txp-content-file_uploadtest.html, based on some previous incarnation of the code in the drop zone branch.

I would be fine with just having the progress bar (`progress` element, that is easy to style anyway).

**Comment by bloatware** [comment]
> I was in no way negating the work you've done here. It is great and very welcome (and very needed). Just time is against us for shipping 4.7!

I would never take it this way, no worries :-) But the more we wait for merging, the less it's automatic.

> Just fwiw, I had made a WIP-attempt at styling a preview box, based on some previous incarnation of the code in the drop zone branch.

Nice! This box must be themable anyway, so we don't have to take a "core" decision. The latest branch introduces à la mustache templates (see line 2020 of `textpattern.js`) that can already be used by theme authors, but we have yet to nail the mechanism in 4.8. Which means you'll probably need to update your theme again.

**Comment by jpdupont** [comment]
Hello everyone ... It would be a pity if this feature (very long waited), even basic was not included in version 4.7 of our dear Textpattern, for files and images ...: - (((

**Comment by philwareham** [comment]
@jpdupont I think we've already established that this (in its basic form) **will** now be in 4.7 - so don't worry! 😀

**Comment by jpdupont** [comment]
Thanks everyone :-)

**Comment by philwareham** [comment]
If this issue specifically is fixed then please close. I'd like to open a general 'multi upload branch' issue so we can track any niggles that need fixing prior to branch merge. Cheers!

**Comment by bloatware** [comment]
Can't think of anything better atm, closing.
